core/modal_models/modes.c

```diff
@@ -335,6 +335,7 @@ void _lf_handle_mode_shutdown_reactions(
  * @param reset_data_size
  * @param timer_triggers Array of pointers to timer triggers.
  * @param timer_triggers_size
+ * FIXME: How does modal models work with LET?  
```

Comment:
Thanks for comment. I am not sure if I understand you correctly. But what about the scenario where we have a LET reaction execute at the same tag as a "mode switching" reaction within the same modal reactor `R1`. What will happen is that the mode change will actually take place while the LET reaction is executing. The mode change is done from `_lf_next_locked` which is called by the last worker advancing time. I don`t think this causes a race condition but I am not sure. We will never "be in 2 modes at the same time" since any new reaction in `R1` will be blocked as an "interrupting reaction" until the LET reaction finishes. 

The question is, is it visible to the LET reaction that we actually have  done the mode  change already? The only thing the LET reaction is allowed to do is schedule actions and edit state variables. Are those "mode-local"?

Reply:
Good question.  I'm not sure.  One option might be to ensure that the mode change acquires the same mutex lock that enforces mutual exclusion of LET reactions and interrupting reactions. In that case, the mode change can't actually complete until the LET reaction completes.

Reply:
That is not a bad idea. That is enforced by the global mutex and the "event_q_changed" condition variable. This is probably the least "surprising" way of implementing this.

Reply:
I don't think we want any reactions holding the global mutex while executing.